{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tacggcaggttccgcctcttagacgccgggaggctcggccggggcgtcggcggccttccctcctgcgaggt......tcgtttctggggttgctcttcgcgctagtgtaccaggacgacctcaagcactggcggcggccctagtgagagccgtacctgctcgacatgttc\n",
      "Complemented sequence copied to clipboard.\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "import pyperclip\n",
    "# Original DNA sequence\n",
    "original_sequence = \"atgccgtccaaggcggagaatctgcggccctccgagccggccccgcagccgccggaagggaggacgctcca......agcaaagaccccaacgagaagcgcgatcacatggtcctgctggagttcgtgaccgccgccgggatcactctcggcatggacgagctgtacaag\"\n",
    "\n",
    "# Convert sequence to Biopython Seq object\n",
    "seq = Seq(original_sequence)\n",
    "\n",
    "# Complement the sequence\n",
    "complement_seq = seq.complement()\n",
    "\n",
    "\n",
    "# Convert the complement sequence to lowercase\n",
    "complement_seq_lower = complement_seq.lower()\n",
    "\n",
    "print(complement_seq_lower)\n",
    "pyperclip.copy(str(complement_seq_lower))\n",
    "print(\"Complemented sequence copied to clipboard.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 491,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tabulate import tabulate\n",
    "\n",
    "def print_primers_table(inserts):\n",
    "    table_data = []\n",
    "\n",
    "    for insert, data in inserts.items():\n",
    "        fw_primer = data.get(\"FW_primer\", [\"Not found\"])\n",
    "        rv_primer = data.get(\"RV_primer\", [\"Not found\"])\n",
    "        table_data.append([insert, \"\", \"Forward Primer\", *fw_primer])\n",
    "        table_data.append([\"\", \"\", \"Reverse Primer\", *rv_primer])\n",
    "        table_data.append([\"\", \"\", \"\", \"\", \"\"])  # Empty row for visual separation\n",
    "\n",
    "    headers = [\"Insert\", \"\", \"Primer Type\", \"Sequence 5'->3'\", \"Tm\", \"GC Content\", \"Start Position\"]\n",
    "    print(tabulate(table_data, headers=headers, tablefmt=\"grid\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 492,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tabulate import tabulate\n",
    "import pandas as pd\n",
    "def primers_to_df(inserts, file_path,format=\"html\"):\n",
    "    headers = [\"Insert\", \"\", \"Primer Type\", \"Sequence 5'->3'\", \"Tm\", \"GC Content\", \"Start Position\"]\n",
    "    table_data = []\n",
    "\n",
    "    for insert, data in inserts.items():\n",
    "        fw_primer = data.get(\"FW_primer\", [\"Not found\"])\n",
    "        rv_primer = data.get(\"RV_primer\", [\"Not found\"])\n",
    "        table_data.append([insert, \"\", \"Forward Primer\", *fw_primer])\n",
    "        table_data.append([\"\", \"\", \"Reverse Primer\", *rv_primer])\n",
    "    df = pd.DataFrame(table_data, columns=headers)\n",
    "    df\n",
    "    #with open(file_path, 'w') as file:\n",
    "    #    file.write(tabulate(table_data, headers=headers, tablefmt=format))\n",
    "    return df\n",
    "# Example usage:\n",
    "# save_primers_table_to_file(inserts, 'primers_table.txt')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 493,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Backbone {'Sequence': Seq('atatgattattacatttatcaaaagagttgatgaagttatacttggatgtagat...gat'), 'FW_primer': None, 'RV_primer': None}\n",
      "*OPCM {'Sequence': Seq('ttacttgtacagctcgtccatgccgccggtggagtggcggccctcggcgcgttc...tag'), 'FW_primer': None, 'RV_primer': None}\n",
      "PKD2L1 {'Sequence': Seq('tcatttcaccgaagccctctgtatcccaagataagacgacaccaaacaaatgtc...aca'), 'FW_primer': None, 'RV_primer': None}\n",
      "INPP5E {'Sequence': Seq('atgccgtccaaggcggagaatctgcggccctccgagccggccccgcagccgccc...aag'), 'FW_primer': None, 'RV_primer': None}\n"
     ]
    }
   ],
   "source": [
    "def read_seqs(file_path):\n",
    "\n",
    "    # Dictionary to store the sequences\n",
    "    sequences = {}\n",
    "\n",
    "    # Temporary variables for keys and values\n",
    "    current_key = None\n",
    "    current_sequence = Seq(\"\")\n",
    "\n",
    "    # Open and read the file line by line\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('#'):\n",
    "                pass\n",
    "            else:\n",
    "                if line.startswith('>'):  # New sequence name detected\n",
    "                    # Save the previous sequence if it exists\n",
    "                    if current_key is not None:\n",
    "                        \n",
    "                        sequences[current_key] = {\"Sequence\":Seq(\"\").join(current_sequence),\"FW_primer\":None,\"RV_primer\":None}\n",
    "                    \n",
    "                    # Reset the current sequence and set the new key\n",
    "                    current_key = line[1:].strip()  # Remove '>' and trim any whitespace\n",
    "                    current_sequence = []\n",
    "                else:\n",
    "                    # Add this line to the current sequence, removing unwanted characters like dots\n",
    "                    current_sequence.append(line.strip())\n",
    "\n",
    "    # Save the last sequence\n",
    "    if current_key is not None:\n",
    "        sequences[current_key] = {\"Sequence\":Seq(\"\").join(current_sequence),\"FW_primer\":None,\"RV_primer\":None}\n",
    "\n",
    "    # Print the result\n",
    "    for key, value in sequences.items():\n",
    "        if '*' in key:\n",
    "            sequences[key][\"Sequence\"] = sequences[key][\"Sequence\"].reverse_complement()\n",
    "          \n",
    "    return sequences\n",
    "inserts = read_seqs('/Users/quillan/Documents/Lab/Thesis/Random stuff/primerstuff/QF_Pkd2l1_Inpp5e.md')\n",
    "for k,v in inserts.items():\n",
    "    print(k,v)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 494,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "\n",
    "def find_primer(sequence, max_primer_start_pos, window_size, max_primer_length, m_t, min_gc):\n",
    "    \"\"\"\n",
    "    Find primers by widening a window then sliding it \n",
    "    \"\"\"\n",
    "    for i in range(max_primer_start_pos):\n",
    "        \n",
    "        for pl in range(window_size, max_primer_length):\n",
    "            primer = sequence[i:i+pl]\n",
    "            primer_tm = mt.Tm_GC(primer)\n",
    "            gc_content = gc_fraction(primer)\n",
    "            \n",
    "            if primer_tm > m_t and gc_content > min_gc :\n",
    "                \n",
    "                if i != 0:\n",
    "                    primer = sequence[0:i]+\"-\"+primer\n",
    "                  \n",
    "                    return  [primer, primer_tm, gc_content, i] \n",
    "                \n",
    "                return  [primer, primer_tm, gc_content, i] # Exit the function when primer found\n",
    "            \n",
    "\n",
    "    return [\"Not found\",\"Not found\",\"Not found\"]\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 495,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_primers(inserts):\n",
    "    window_size = 10\n",
    "    max_primer_length = 30\n",
    "    max_primer_start_pos = 10\n",
    "    min_gc = 0.5\n",
    "    m_t = 55\n",
    "    for insert,data in inserts.items():\n",
    "        data[\"FW_primer\"] = None\n",
    "        data[\"RV_primer\"] = None\n",
    "\n",
    "        # fw primer\n",
    "        gc_adj = 0\n",
    "        m_t_adj = 0\n",
    "        data[\"FW_primer\"] = find_primer(data['Sequence'],max_primer_start_pos,window_size,max_primer_length,m_t,min_gc)\n",
    "        \n",
    "        while \"Not found\" in data[\"FW_primer\"]:\n",
    "            data[\"FW_primer\"] = find_primer(data['Sequence'],max_primer_start_pos,window_size,max_primer_length,m_t-m_t_adj,min_gc-gc_adj)\n",
    "            \n",
    "            \n",
    "            if min_gc-gc_adj > 0.1:\n",
    "                gc_adj += 0.1\n",
    "                \n",
    "            if m_t-m_t_adj > 45:\n",
    "                m_t_adj += 0.1\n",
    "                \n",
    "            elif data[\"FW_primer\"][0] == \"Not found\":\n",
    "                data[\"FW_primer\"] = [\"Not Found.\"]\n",
    "\n",
    "        # rv primer\n",
    "        gc_adj = 0\n",
    "        m_t_adj = 0\n",
    "        data[\"RV_primer\"] = find_primer(data['Sequence'].reverse_complement(),max_primer_start_pos,window_size,max_primer_length,m_t,min_gc)\n",
    "        while \"Not found\" in data[\"RV_primer\"]:\n",
    "            data[\"RV_primer\"] = find_primer(data['Sequence'].reverse_complement(),max_primer_start_pos,window_size,max_primer_length,m_t-m_t_adj,min_gc-gc_adj)\n",
    "                \n",
    "            if min_gc-gc_adj > 0.1:\n",
    "                gc_adj += 0.1\n",
    "            if m_t-m_t_adj > 45:\n",
    "                m_t_adj += 0.1\n",
    "            elif data[\"RV_primer\"][0] == \"Not found\":\n",
    "                data[\"RV_primer\"] = [\"Not Found.\"]\n",
    "    return inserts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 542,
   "metadata": {},
   "outputs": [],
   "source": [
    "homology_seq_length = 10\n",
    "def add_homology(inserts,hom_length=10):\n",
    "    hom_length+=1\n",
    "    insert_names = list(inserts.keys())\n",
    "    for i,name in enumerate(insert_names):\n",
    "        \n",
    "        if i == len(insert_names)-1:\n",
    "            hom_seq = inserts[insert_names[0]][\"Sequence\"].complement()[:homology_seq_length]\n",
    "            inserts[insert_names[i]][\"RV_primer\"][0] = f'{hom_seq}-{inserts[insert_names[i]][\"RV_primer\"][0]}'\n",
    "        else:\n",
    "            hom_seq = inserts[insert_names[i+1]][\"Sequence\"].complement()[:homology_seq_length]\n",
    "            inserts[insert_names[i]][\"RV_primer\"][0] = f'{hom_seq}-{inserts[insert_names[i]][\"RV_primer\"][0]}'\n",
    "\n",
    "        if i == 0:\n",
    "            hom_seq = inserts[insert_names[-1]][\"Sequence\"][-homology_seq_length:]\n",
    "            inserts[insert_names[i]][\"FW_primer\"][0] = f'{hom_seq}-{inserts[insert_names[i]][\"FW_primer\"][0]}'\n",
    "        else:\n",
    "            hom_seq = inserts[insert_names[i-1]][\"Sequence\"][-homology_seq_length:]\n",
    "            inserts[insert_names[i]][\"FW_primer\"][0] = f'{hom_seq }-{inserts[insert_names[i]][\"FW_primer\"][0]}'\n",
    "            \n",
    "    return inserts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 543,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "import pyperclip\n",
    "\n",
    "# find primers\n",
    "inserts = get_primers(inserts)\n",
    "\n",
    "# append homology sequence overhang\n",
    "inserts = add_homology(inserts)\n",
    "df = primers_to_df(inserts,\"Primers_SLIC.html\")\n",
    "\n",
    "html_df = df.to_html()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 548,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INPP5E\n",
      "<p>----------tacggcaggttccgcctcttagacgccgggaggctcggccggggcgtcggcggga...taccaggacgacctcaagcactggcggcggccctagtgagagccgtacctgctcgacatgttc<span class=\"BackboneINPP5E\">tatactaata</span></p>\n"
     ]
    }
   ],
   "source": [
    "def export(inserts):\n",
    "    insert_names = list(inserts.keys(),homology_seq_length)\n",
    "    for id, name in enumerate(insert_names):\n",
    "\n",
    "        # append primers to sequence for visualization purposes\n",
    "        if id == 0:\n",
    "            poi = insert_names[id]  # part of interest\n",
    "            homology =  \"\".join(sorted([poi,insert_names[-1]])).replace(\"*\",\"rev\")\n",
    "            inserts[poi][\"FW_sequence_woverhang\"]=f'<p><span class=\"{homology}\">{inserts[poi][\"FW_primer\"][0].split(\"-\")[0]}</span>{inserts[poi][\"Sequence\"]}{\"-\"*homology_seq_length}</p>'\n",
    "        \n",
    "        else:\n",
    "            poi = insert_names[id]  # part of interest\n",
    "            homology =  \"\".join(sorted([poi,insert_names[id-1]])).replace(\"*\",\"rev\")\n",
    "            inserts[poi][\"FW_sequence_woverhang\"]=f'<p><span class=\"{homology}\">{inserts[poi][\"FW_primer\"][0].split(\"-\")[0]}</span>{inserts[poi][\"Sequence\"]}{\"-\"*homology_seq_length}</p>'\n",
    "            \n",
    "        if id == len(insert_names)-1:\n",
    "            poi = insert_names[id]  # part of interest\n",
    "            print(poi)\n",
    "            homology =  \"\".join(sorted([poi,insert_names[0]])).replace(\"*\",\"rev\")\n",
    "            inserts[poi][\"RV_sequence_woverhang\"]=f'<p>{\"-\"*homology_seq_length}{inserts[poi][\"Sequence\"].complement()}<span class=\"{homology}\">{inserts[poi][\"RV_primer\"][0].split(\"-\")[0]}</span></p>'\n",
    "        else:\n",
    "            poi = insert_names[id]  # part of interest\n",
    "            homology =  \"\".join(sorted([poi,insert_names[id+1]])).replace(\"*\",\"rev\")\n",
    "            inserts[poi][\"RV_sequence_woverhang\"]=f'<p>{\"-\"*homology_seq_length}{inserts[poi][\"Sequence\"].complement()}<span class=\"{homology}\">{inserts[poi][\"RV_primer\"][0].split(\"-\")[0]}</span></p>'\n",
    "\n",
    "print(inserts[\"INPP5E\"][\"RV_sequence_woverhang\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "metadata": {},
   "outputs": [],
   "source": [
    "js = \"\"\"\n",
    "<script>\n",
    "// Define a function to generate a random color\n",
    "function getRandomColor() {\n",
    "  const letters = '0123456789ABCDEF';\n",
    "  let color = '#';\n",
    "  for (let i = 0; i < 6; i++) {\n",
    "    color += letters[Math.floor(Math.random() * 16)];\n",
    "  }\n",
    "  return color;\n",
    "}\n",
    "\n",
    "// Get all elements in the DOM\n",
    "const allElements = document.querySelectorAll('*');\n",
    "\n",
    "// Create a Set to store unique class names\n",
    "const uniqueClasses = new Set();\n",
    "\n",
    "// Iterate through all elements to collect unique class names\n",
    "allElements.forEach(element => {\n",
    "  const classes = element.classList;\n",
    "  classes.forEach(className => {\n",
    "    if (className != 'dataframe'){\n",
    "    uniqueClasses.add(className);\n",
    "    }\n",
    "  });\n",
    "});\n",
    "\n",
    "// Assign a random color to all elements sharing each class\n",
    "uniqueClasses.forEach(className => {\n",
    "  const elementsWithClass = document.querySelectorAll('.' + className);\n",
    "  const randomColor = getRandomColor();\n",
    "  elementsWithClass.forEach(element => {\n",
    "    element.style.backgroundColor = randomColor;\n",
    "  });\n",
    "});\n",
    "</script>\n",
    "<style>\n",
    "p {\n",
    "    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;\n",
    "    font-size: 14px;\n",
    "    line-height: 1.5;\n",
    "    padding: 10px;\n",
    "    background-color: #f4f4f4;\n",
    "    border: 1px solid #ddd;\n",
    "    border-radius: 5px;\n",
    "    overflow-x: auto; /* Enable horizontal scrolling if needed */\n",
    "    white-space: pre-wrap; /* Preserve line breaks */\n",
    "}\n",
    "/* Reset default table styles */\n",
    "table {\n",
    "  border-collapse: collapse;\n",
    "  width: 100%;\n",
    "}\n",
    "\n",
    "/* Table header styles */\n",
    "thead {\n",
    "  background-color: #f2f2f2 !important;\n",
    "}\n",
    "\n",
    "/* Table cell styles */\n",
    "td, th {\n",
    "  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;\n",
    "  border: 1px solid #dddddd;\n",
    "  text-align: left;\n",
    "  padding: 8px;\n",
    "}\n",
    "\n",
    "/* Alternate row color */\n",
    "\n",
    "\n",
    "/* Hover effect */\n",
    "tbody tr:hover {\n",
    "  background-color: #f2f2f2 !important;\n",
    "}\n",
    "\n",
    "/* Modern table style */\n",
    ".modern-table {\n",
    "  border-radius: 8px;\n",
    "  overflow: hidden;\n",
    "  box-shadow: 0 0 20px rgba(0, 0, 0, 0.1);\n",
    "}\n",
    "h2{\n",
    "font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "</style>\n",
    "\"\"\"\n",
    "Title=\"\"\"\n",
    "<h2>SLIC helper results</h2>\n",
    "Below are the primers found in [homology-ext-binding] format. The Tm is for the binding sequence only\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 552,
   "metadata": {},
   "outputs": [],
   "source": [
    "outname = \"output\"\n",
    "with open(outname+\".html\", 'w') as file:\n",
    "        file.write(Title)\n",
    "        file.write(html_df)\n",
    "        for insert,data in inserts.items():\n",
    "                file.write(f'<h2>{insert}</h2>')\n",
    "                file.write(f'{data[\"FW_sequence_woverhang\"]}')\n",
    "                file.write(f'{data[\"RV_sequence_woverhang\"]}')\n",
    "        file.write(f'<h2>Final construct</h2>')\n",
    "        file.write(f'<p>{str(finalconstruct)}</p>')     \n",
    "        file.write(js)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 551,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create final construct\n",
    "def create_final_construct(inserts):\n",
    "    finalconstruct = Seq(\"\")\n",
    "    outname = \"\"\n",
    "    for insert in inserts:\n",
    "        outname += str(insert)\n",
    "        finalconstruct += inserts[insert][\"Sequence\"]\n",
    "\n",
    "    with open(outname+\".txt\", 'w') as file:\n",
    "            file.write(str(finalconstruct))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
